### PR TITLE
make each platform build independently

### PR DESCRIPTION
**Pull request by jhheider**
moves matrices up a level. depends on #418

The secret sauce is `scripts/get-platform.ts` which generates a bunch of values for us, in a centralized fashion. Then we use them inline. Of particular coolness is using it and `fromJson()` to generate a `matrix:` for our testing strategy.

### CODE REVIEW

File: scripts/get-platform.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      break
+  case "linux+aarch64":
+      rv = { os: "[self-hosted, linux, ARM64]", container: "~" }
+      break
+  case "linux+x86-64":
+    rv = { os: await getLinuxSize(), container: "debian:buster-slim" }
+    break
+  default:
+    panic(`Invalid platform description: ${platform}`)
+}
```
Comment (jhheider) [inline scripts/get-platform.ts:36]: Pretty glad to be concentrating matrix info in one place.

**Comment by jhheider** [comment]
Short of an ops change needed on the macOS X64 builder, I think this is working.

File: scripts/get-platform.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    testMatrix = [
+      { os, container: undefined },
+      { os: buildOs, container: undefined },
+      { os, container: "ghcr.io/teaxyz/infuser:latest" },
+      { os, container: "debian:buster-slim" },
+    ]
+    break
+  default:
+    panic(`Invalid platform description: ${platform}`)
+}
```
Comment (jhheider) [inline scripts/get-platform.ts:48]: this is pretty cool, ngl.

Comment (jhheider) [inline scripts/get-platform.ts:48]: this is pretty cool, ngl.

File: .github/workflows/bottle.yml [hunk 3]
```diff
@@ -81,14 +84,14 @@ jobs:
...
-        if: matrix.platform.name == 'darwin+aarch64' || matrix.platform.name == 'darwin+x86-64'
+        if: startsWith(inputs.platform, 'darwin+')
         with:
           p12-file-base64: ${{ secrets.APPLE_CERTIFICATE_P12 }}
           p12-password: ${{ secrets.APPLE_CERTIFICATE_P12_PASSWORD }}
 
       # Codesign libs and bins
       - name: Codesign package
-        if: matrix.platform.name == 'darwin+aarch64' || matrix.platform.name == 'darwin+x86-64'
+        if: startsWith(inputs.platform, 'darwin+')
```
Comment (mxcl) [inline .github/workflows/bottle.yml:94]: ur getn gud at this

File: .github/workflows/bottle.yml [hunk 4]
```diff
@@ -7,24 +7,31 @@ on:
...
+      - uses: teaxyz/setup@v0
+      - run: scripts/get-platform.ts ${{ inputs.projects }}
+        id: platform
+        env:
+          PLATFORM: ${{ inputs.platform }}
+
   bottle:
-    runs-on: ${{ matrix.platform.os }}
+    needs: [get-platform]
+    runs-on: ${{ fromJson(needs.get-platform.outputs.os) }}
```
Comment (mxcl) [inline .github/workflows/bottle.yml:31]: sweet

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -6,38 +6,35 @@ on:
...
 
 jobs:
-  check-core-sizes:
+  get-platform:
     runs-on: ubuntu-latest
     outputs:
-      gha-linux-build-size: ${{ env.GHA_LINUX_BUILD_SIZE }}
+      os: ${{ steps.platform.outputs.os }}
+      build-os: ${{ steps.platform.outputs.build-os }}
+      testMatrix: ${{ steps.platform.outputs.testMatrix }}
```
Comment (mxcl) [inline .github/workflows/build.yml:23]: `test-matrix` for consistency

File: deno.lock [hunk 6]
```diff
@@ -0,0 +1,282 @@
+{
```
Comment (mxcl) [inline deno.lock:1]: I added this to `.gitignore` in latest merge. Let’s not commit until v1.

**Review by mxcl — APPROVED** [review]
Seems right. Let’s find out.

File: .github/workflows/bottle.yml (hunk 3 above)
Comment (jhheider) [reply .github/workflows/bottle.yml:94]: Practice.

**Pull request merged**